nixos/modules/services/networking/iwd.nix

```diff
@@ -4,16 +4,142 @@ with lib;
 
 let
   cfg = config.networking.wireless.iwd;
+
+  encodeFileName = pkgs.writeScript "encode" ''
+    #! ${pkgs.runtimeShell} -e
+    echo "=$(printf "$1" | ${pkgs.unixtools.xxd}/bin/xxd -pu)"
+  '';
+
+  # If a network consists of alphanumeric chars and `_` and `-`, the network will be stored
+  # as `/var/lib/iwd/<ssid>.<ext>`. In any other case a hex-encoded version of the SSID will
+  # be stored with an `=` as prefix.
+  encodeIfNeeded = file: ext: let
+    allowedChars = stringToCharacters "01234556789abcdefghijklmnopqrstuvwxyz-_ ";
+    skipEncoding = all (char: elem (toLower char) allowedChars) (stringToCharacters file);
+  in
+    "/var/lib/iwd/${if skipEncoding then file else "$(${encodeFileName} \"${file}\")"}.${ext}";
+
+  # Basically `types.str`, but the `check` function validates the string-length as well.
+  fixedStringLength = min: max: mkOptionType {
+    name = "str-with-length";
+    description = if min == max
+      then "string with exactly ${toString min} chars."
+      else "string which must have at least ${toString min} and at most ${toString max} chars.";
+    inherit (types.str) merge;
+    check = x: isString x && stringLength x >= min && stringLength x <= max;
```

Comment:
Since this type would be useful for many other modules, I suggest adding it to `lib/types.nix`. I'd also name it `lengthCheckedString` instead. Note that `name` should also be set to what the type variable is named when doing this.